distributed/comm/asyncio_tcp.py

```diff
@@ -0,0 +1,969 @@
+        self._transport = None
+        # Is the protocol paused?
+        self._paused = False
+        # If the protocol is paused, this holds a future to wait on until it's
+        # unpaused.
+        self._drain_waiter: asyncio.Future | None = None
+        # A future for waiting until the protocol is actually closed
+        self._closed_waiter = self._loop.create_future()
+
+        # In the interest of reducing the number of `recv` calls, we always
+        # want to provide the opportunity to read `min_read_size` bytes from
+        # the socket (since memcpy is much faster than recv). Each read event
+        # may read into either a default buffer (of size `min_read_size`), or
+        # directly into one of the message frames (if the frame size is >
+        # `min_read_size`).
+
+        # Per-message state
+        self._using_default_buffer = True
+
+        self._default_len = max(min_read_size, 16)  # need at least 16 bytes of buffer
+        self._default_buffer = memoryview(bytearray(self._default_len))
+        # Index in default_buffer pointing to the first unparsed byte
+        self._default_start = 0
+        # Index in default_buffer pointing to the last written byte
+        self._default_end = 0
+
+        # Each message is composed of one or more frames, these attributes
+        # are filled in as the message is parsed, and cleared once a message
+        # is fully parsed.
+        self._nframes: int | None = None
+        self._frame_lengths: list[int] | None = None
+        self._frames: list[memoryview] | None = None
+        self._frame_index: int | None = None  # current frame to parse
+        self._frame_nbytes_needed: int = 0  # nbytes left for parsing current frame
+
+    @property
+    def local_addr(self):
+        if self.is_closed:
+            return "<closed>"
+        sockname = self._transport.get_extra_info("sockname")
+        if sockname is not None:
+            return unparse_host_port(*sockname[:2])
+        return "<unknown>"
+
+    @property
+    def peer_addr(self):
+        if self.is_closed:
+            return "<closed>"
+        peername = self._transport.get_extra_info("peername")
+        if peername is not None:
+            return unparse_host_port(*peername[:2])
+        return "<unknown>"
+
+    @property
+    def is_closed(self):
+        return self._transport is None
+
+    def _abort(self):
+        if not self.is_closed:
+            self._transport, transport = None, self._transport
+            transport.abort()
+
+    def _close_from_finalizer(self, comm_repr):
+        if not self.is_closed:
+            logger.warning(f"Closing dangling comm `{comm_repr}`")
+            try:
+                self._abort()
+            except RuntimeError:
+                # This happens if the event loop is already closed
+                pass
+
+    async def _close(self):
+        if not self.is_closed:
+            self._transport, transport = None, self._transport
```

Comment:
A swap like this actually isn't thread safe. This is to ensure that regardless of failure calling `_close` again is a no-op (since `self._transport` will be `None`).